src/object/cli_obj.c

```diff
@@ -3258,38 +3252,50 @@ obj_shard_list_obj_cb(struct shard_auxi_args *shard_auxi,
 	daos_obj_list_t		*obj_arg = dc_task_get_args(obj_auxi->obj_task);
 	char			*ptr = obj_arg->sgl->sg_iovs[0].iov_buf;
 	daos_key_desc_t		*kds = obj_arg->kds;
+	int			i;
 	int			rc;
 
 	shard_arg = container_of(shard_auxi, struct shard_list_args, la_auxi);
 	rc = obj_enum_iterate(shard_arg->la_kds, shard_arg->la_sgl,
 			      shard_arg->la_nr, OBJ_ITER_RECX,
 			      obj_enum_shard_process_cb, shard_auxi);
 	ptr += iter_arg->merge_sgl_off;
-	if (ptr != shard_arg->la_sgl->sg_iovs[0].iov_buf)
-		memmove(ptr, shard_arg->la_sgl->sg_iovs[0].iov_buf,
-			shard_arg->la_sgl->sg_iovs[0].iov_len);
+	memcpy(ptr, shard_arg->la_sgl->sg_iovs[0].iov_buf,
+	       shard_arg->la_sgl->sg_iovs[0].iov_len);
 	obj_arg->sgl->sg_iovs[0].iov_len +=
-			shard_arg->la_sgl->sg_iovs[0].iov_len;
+		shard_arg->la_sgl->sg_iovs[0].iov_len;
 	iter_arg->merge_sgl_off += shard_arg->la_sgl->sg_iovs[0].iov_len;
 
 	kds += iter_arg->merge_nr;
-	if (kds != shard_arg->la_kds) {
-		int i;
-
-		for (i = 0; i < shard_arg->la_nr; i++)
-			kds[i] = shard_arg->la_kds[i];
-	}
+	for (i = 0; i < shard_arg->la_nr; i++)
+		kds[i] = shard_arg->la_kds[i];
 	iter_arg->merge_nr += shard_arg->la_nr;
 
 	D_DEBUG(DB_TRACE, "merge_nr %d/"DF_U64"\n", iter_arg->merge_nr,
 		obj_arg->sgl->sg_iovs[0].iov_len);
 	return rc;
 }
 
+uint64_t
+enum_hkey_gen(struct dc_object *obj, daos_key_t *key)
+{
+	if (daos_is_dkey_uint64(obj->cob_md.omd_id)) {
```

Comment:
see my comment below.   If we init hkey to 0, we can memcpy anything <= sizeof(uint64_t) regardless of whether it's an integer key.  I presume collisions don't matter, right?   One reason we use two hashes and inline value on server is to try to avoid any hash collisions because the btree doesn't handle them.

Reply:
I think client has to use the same order as VOS key to sort,  otherwise we might see the duplication, so this has to be the same as btr_hkey_gen &ktr_hkey_gen.
Ah, I may need some fixes here. Thanks.